Fixed Improvements to error messaging issue #1091

devkashan1 opened the pull request:
The details of the implementation for this problem that I have fixed in the project as I briefed above:
I have added a new file where we can manage any error messages at this place `src\libs\ErrorMessage\messages.js.`
I have also added another file where I created a function to fetch the appropriate message via type and error code. The path is `src\libs\ErrorMessage\index.js.`

After these, I have modified a file `src\libs\actions\Session.js` where we call the `getErrorMessage` function.
Line #11: `import {getErrorMessage} from '../ErrorMessage';`
Line #162: `Onyx.merge(ONYXKEYS.ACCOUNT, {error: getErrorMessage('Login',error.message)});`
Line #169: `Onyx.merge(ONYXKEYS.ACCOUNT, {error: getErrorMessage('Login',error.message), loading: false});`

That's all.

devkashan1 committed:
Fixed Improvements to error messaging issue #1091

devkashan1 commented:
I have read the CLA Document and I hereby sign the CLA

devkashan1 commented on src/libs/ErrorMessage/index.js:
```diff
@@ -0,0 +1,21 @@
+ */
+function getErrorMessage(type, error) {
+  const code = error.split(' ')[0];
+  if (!_.isEmpty(code)) {
```
It was set via prettify formatter.

devkashan1 reviewed (commented):
I have reviewed everything. It was set via prettify formatter.

deetergp reviewed (changes requested):
Please address the linter failures. If you look at the rest of the code base you'll see that we use four-space indentation everywhere. Please do the same for src/libs/ErrorMessage/messages.js

devkashan1 committed:
Linter fixes, added 2 spaces.

devkashan1 committed:
Fixed a couple of spaces

devkashan1 committed:
Fixed a few move lines

devkashan1 committed:
Fixed spaces issues.

devkashan1 committed:
property shorthand fixed

devkashan1 committed:
object-shorthand fixed

devkashan1 commented:
Thank for letting me. I've fixed linter failures. Can you please review it now.

devkashan1 committed:
fixed the error messages

devkashan1 committed:
removed type

yuwenmemon commented on src/libs/ErrorMessage/errorMessages.js:
```diff
@@ -0,0 +1,31 @@
+const erroMessages = [
```
Typo

devkashan1 replied:
Got it. Thanks

devkashan1 replied:
Done, You can verify please.

yuwenmemon commented on src/libs/ErrorMessage/index.js:
```diff
@@ -0,0 +1,20 @@
+import errorMessages from './errorMessages';
+
+/**
+ * It's generic function to print any messages.
```
Let's remove the generic aspect of this. It's over-engineering. Right now we just need to handle error messages better in the context of `Authenticate`.

devkashan1 replied:
Sure!

devkashan1 replied:
Done, You can verify please.

devkashan1 commented:
Here is the details of the changes.
Replace a file name `src\libs\ErrorMessage\messages.js.` to `src\libs\ErrorMessage\errorMessages.js.` due to avoid misunderstanding of naming.
I had removed the `type` parameter that was considering as generic but now we can pass only parameter `getErrorMessage(error.message)`

Also, I have modified a file src\libs\actions\Session.js where we call the getErrorMessage function.
Line #162: Onyx.merge(ONYXKEYS.ACCOUNT, {error: getErrorMessage(error.message)});
Line #169: Onyx.merge(ONYXKEYS.ACCOUNT, {error: getErrorMessage(error.message), loading: false});

After that, I have changed this code for calling the error with one parameter and changed the `message `obj to `errorMessage`  in `src\libs\ErrorMessage\index.js`
```
function getErrorMessage(error) {
    const code = error.split(' ')[0];
    if (!_.isEmpty(code)) {
        const errorMessage = _.filter(errorMessages, {errorCode: +code.trim()});
        if (!_.isEmpty(errorMessage)) {
            return errorMessage[0].errorMessage;
        }
    }
    return error;
}
export default getErrorMessage;
```

Please let me know if you have any questions.

devkashan1 committed:
typo fixed and changed the description of function

deetergp commented on src/libs/ErrorMessage/index.js:
```diff
@@ -2,7 +2,7 @@ import _ from 'lodash';
 
 /**
- * It's generic function to print any messages.
+ * It's function to print handle error messages in the context of Authenticate.
```
Typo? `print handle error messages` is awkward. Probably should be either/or.

devkashan1 replied:
@deetergp Okay, I think, It will be better. `This function is handling error messages for the Authentication.` else can you advise on it?

deetergp replied:
At the end of the day, shouldn't your naming explain the "what" and your docs and comments explain the "why"? Like shouldn't function be called `showFriendlyErrorMessage` and the doc comment should briefly summarize the reason we need the function like "Display a user-friendly message in the UI when the API returns an authentication error".

devkashan1 replied:
@deetergp  Okay, Should we change with this `Display a error message in the UI when the API returns an authentication error.`?
If it works, please let me know. I'll change it. Thanks

yuwenmemon replied:
@devkashan1 I meant to move this logic to the context of `Authenticate` (in src/libs/Api.js) since that's the only spot where it's relevant.

devkashan1 replied:
@yuwenmemon Thanks for the clarification. I just have pushed the changes. You can verify now.

devkashan1 committed:
Moved & changed the function in Session.js

devkashan1 committed:
moved the function to API.js

devkashan1 committed:
Linter fixes

yuwenmemon commented on src/libs/API.js:
```diff
@@ -481,6 +482,23 @@ function SetPassword(parameters) {
+ * @param {String} error
+ * @returns {String}
+ */
+function GetSignInErrorMessage(error) {
```
Ah what I meant was that you should probably place this in `Authenticate`. I also think you don't need a separate authenticationErrorMessages file either. It's overkill for what we're trying to achieve IMO

devkashan1 replied:
What I understood where you mentioned place in `Authenticate`. This means in `src\libs\actions\Session.js` file. So, This function `GetSignInErrorMessage ` should be in this `src\libs\actions\Session.js` file, and no need separate file for `authenticationErrorMessages`. If you confirm me so I can proceed the changes.

devkashan1 replied:
@yuwenmemon Could you please confirm so I can push and close it?

yuwenmemon commented:
Closing this PR as a solution was already merged
